### PR TITLE
Support reference count in varlen type

### PR DESCRIPTION
**Pull request by jessesleeping**
In this PR, we implemented the following features:
- Support reference count in varlen pool. 
- Make the update executor aware of the reference count and perform a shallow copy on varlen column
- Make the GC manager aware of the reference count and reclaim memory in the varlen pool.

#396

### CODE REVIEW

**Comment by jessesleeping** [comment]
Now we can fully support reference count in varlen type. However, it's kind of hacking the type system to support the shallow copy feature we want.

I have profiled the performance gain from this feature with the YCSB micro-benchmark. Here is the [result](https://docs.google.com/spreadsheets/d/1nL0-rUCTE_niGRLvMt3paW1RZruc_Cavblno98sG99o/edit#gid=0).

Basically, we get 10x throughput at 10 thread compared to the original varlen pool. The performance is closed to the YCSB using integer instead. Compared to our old varlen pool, the new one is way too slow and not scalable.

**Comment by jessesleeping** [comment]
Now we can fully support reference count in varlen type. However, it's kind of hacking the type system to support the shallow copy feature we want.

I have profiled the performance gain from this feature with the YCSB micro-benchmark. Here is the [result](https://docs.google.com/spreadsheets/d/1nL0-rUCTE_niGRLvMt3paW1RZruc_Cavblno98sG99o/edit#gid=0).

Basically, we get 10x throughput at 10 thread compared to the original varlen pool. The performance is closed to the YCSB using integer instead. Compared to our old varlen pool, the new one is way too slow and not scalable.

**Comment by jessesleeping** [comment]
Also, I think we need to make shallow copy as the default copy method for varlen type. Currently a simple `GetValue` call to a `ConatinerTuple` will incur 3~4 memory copy of a varlen value.

**Comment by malin1993ml** [comment]
@jessesleeping What exactly are the "original varlen pool", "old varlen pool", and "new one" you're referring to in [this comment](https://github.com/cmu-db/peloton/pull/405#issuecomment-267752155)? Are the "original varlen pool" and the "old varlen pool" the same thing? Also, are your results gotten with OLTPBench or micro-benchmark?

**Comment by jessesleeping** [comment]
Sorry for the confusing wording. "The original" one means the one in our current

**Comment by jessesleeping** [comment]
Sorry for the confusing wordings... I have corrected them in the commend. I tested it with the micro benchmark.

**Comment by malin1993ml** [comment]
Got it. Thanks!

File: src/common/value.cpp [hunk 1]
```diff
@@ -394,6 +394,12 @@ void Value::SerializeTo(SerializeOutput &out) const {
   type_->SerializeTo(*this, out);
 }
 
+// Perform a shallow copy from a serialized varlen value to another serialized varlen value
+// Only support VARCHAR/VARBINARY
+void Value::ShallowCopyTo(char *dest, char *src, const Type::TypeId type_id, bool inlined, VarlenPool *src_pool) {
+  Type::GetInstance(type_id)->DoShallowCopy(dest, src, inlined, src_pool);
+}
```
Comment (malin1993ml) [inline src/common/value.cpp]: Is there a deep copy version for this? Or the Copy() function below will just do the deep copy?

File: src/planner/project_info.cpp [hunk 2]
```diff
@@ -80,30 +82,30 @@ bool ProjectInfo::Evaluate(storage::Tuple *dest, const AbstractTuple *tuple1,
...
+    for (auto dm : direct_map_list_) {
+      // whether left tuple or right tuple ?
+      auto tuple_index = dm.second.first;
+      auto src_col_id = dm.second.second;
+
+      PL_ASSERT(dm.first == dm.second.second);
+      PL_ASSERT(tuple_index == 0);
+      if (tuple_index == 0) {
+        src->CopyColumnTo(dest, src_col_id);
+      }
```
Comment (malin1993ml) [inline src/planner/project_info.cpp]: I don't quite understand the logic here. The comment said "Only copy un-modified value if not in-place update", but it seems that the code suggested if "inplace == false", then we call CopyColumnTo(), which calls the shallow copy function. So should we do or not do a deep copy when it's in-place update? Also, shouldn't there be another branch of logic to do the deep copy on the other condition?

File: src/common/value.cpp (hunk 1 above)
Comment (jessesleeping) [inline src/common/value.cpp]: The Copy() function below is the deep copy one.

File: src/common/value.cpp [hunk 3]
```diff
@@ -404,6 +410,19 @@ Value Value::CastAs(const Type::TypeId type_id) const {
...
+// Access the raw variable length data from a pointer pointed to a tuple storage
+char *Value::GetDataFromStorage(Type::TypeId type_id, char *storage) {
+  switch (type_id) {
+    case Type::VARCHAR:
+    case Type::VARBINARY: {
+      return Type::GetInstance(type_id)->GetData(storage);
+    }
+    default:
+      throw Exception(EXCEPTION_TYPE_INCOMPATIBLE_TYPE,
+                      "Invalid Type for constructor");
```
Comment (jessesleeping) [inline src/common/value.cpp]: Fixed. Thank you for pointing it out.

File: src/planner/project_info.cpp (hunk 2 above)
Comment (malin1993ml) [inline src/planner/project_info.cpp]: Got it. That makes sense.

**Pull request merged**